ms2deepscore/data_generators.py

```diff
@@ -382,71 +376,44 @@ def __len__(self):
         """Denotes the number of batches per epoch"""
         # TODO: this means we don't see all data every epoch, because the last half-empty batch
         #  is omitted. I guess that is expected behavior? --> Yes, with the shuffling in each epoch that seem OK to me (and makes the code easier).
-        return int(self.settings["num_turns"]) * int(np.floor(len(self.inchikey_ids) / self.settings["batch_size"]))
+        return int(self.settings["num_turns"]) * int(np.floor(len(self.labels_df) / self.settings[
+            "batch_size"]))
 
     def __getitem__(self, index):
         """Generate one batch of data"""
         # Go through all indexes
-        indexes_inchkeys = self.indexes[index*self.settings["batch_size"]:(index+1)*self.settings["batch_size"]]
+        indexes_inchikeys = self.indexes[index*self.settings["batch_size"]:(index+1)*self.settings["batch_size"]]
 
         # Select subset of IDs
         # TODO: Filter out function: select_batch_ids
-        inchikey_ids_batch = []
+        inchikeys_batch = []
         same_prob_bins = self.settings["same_prob_bins"]
-        for index in indexes_inchkeys:
-            inchikey_id1 = self.inchikey_ids[index]
-
+        for index in indexes_inchikeys:
+            inchikey1 = self.labels_df.index[index]
             # Randomly pick the desired target score range and pick matching ID
             target_score_range = same_prob_bins[np.random.choice(np.arange(len(same_prob_bins)))]
-            inchikey_id2 = self.__find_match_in_range(inchikey_id1, target_score_range)
+            inchikey2 = self._find_match_in_range(inchikey1, target_score_range)
 
-            inchikey_ids_batch.append((inchikey_id1, inchikey_id2))
+            inchikeys_batch.append((inchikey1, inchikey2))
 
         # Generate data
-        X, y = self.__data_generation(inchikey_ids_batch)
+        X, y = self.__data_generation(inchikeys_batch)
 
         return X, y
 
+    @staticmethod
+    def _data_selection(labels_df, selected_inchikeys):
+        """
+        Select labeled data to generate from based on `selected_inchikeys`
+        """
+        return labels_df.loc[selected_inchikeys, selected_inchikeys]
+
     def on_epoch_end(self):
         """Updates indexes after each epoch"""
-        self.indexes = np.tile(np.arange(len(self.inchikey_ids)), int(self.settings["num_turns"]))
+        self.indexes = np.tile(np.arange(len(self.labels_df)), int(self.settings["num_turns"]))
         if self.settings["shuffle"] == True:
             np.random.shuffle(self.indexes)
 
-    def __find_match_in_range(self, inchikey_id1, target_score_range, max_range=0.4):
```

Comment:
I looked very carefully, but I couldn't find any differences with `DataGeneratorAllSpectrums.__find_match_in_range`. That's correct right?

Reply:
Yes, I think that's correct. I was also thinking of creating a parent class and inherit from there... but then postponed it for now.